Src/Support/GoogleApis.Auth.DotNet4/OAuth2/AuthorizationDialog.cs

```diff
@@ -0,0 +1,47 @@
+﻿using System;
+using System.Collections.Generic;
+using System.ComponentModel;
+using System.Data;
+using System.Drawing;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Windows.Forms;
+using Google.Apis.Auth.OAuth2.Responses;
+
+namespace Google.Apis.Auth.OAuth2
+{
+    public partial class AuthorizationDialog : Form
+    {
+        private AuthorizationCodeResponseUrl responseUrl;
+
+        public AuthorizationCodeResponseUrl ResponseUrl
+        {
+            get
+            {
+                return responseUrl;
+            }
+        }
+        /// <summary> Constructs a new authentication broker dialog. </summary>
+        public AuthorizationDialog(Uri authUri)
+        {
+            InitializeComponent();
+            // Set the browser to the initial uri.
+            authWebBrowser.Navigate(authUri);
+            // Need to track when the results has been received.
+            authWebBrowser.DocumentCompleted += authWebBrowser_DocumentCompleted;
```

Comment:
@LindaLawton I did some desk checking of what would be required to expose the Authorization URI to the user and allow them to use their own in-process browser to handle receiving the response code from Google.
## Setting context of my thought process
The entire task is a sequence that starts with a code request and ends with a response code token. Everything in-between is currently handled entirely by the library. There are a lot of async operations, but the final result is still procedural since we will always await on the response code.
## Current flow
This is my understanding of the current flow, from a higher level, with many ancillary details omitted.

1. AuthorizationCodeInstalledApp.AuthorizeAsync gets the authorization URI from an IAuthorizationCodeFlow instance.
1. AuthorizationCodeInstalledApp.AuthorizeAsync calls (and awaits) ReceiveCodeAsync on ICodeReceiver.
 * This happens to be an instance of EmbeddedBrowserCodeReceiver.
 * The authorization URI is passed as a parameter
1. EmbeddedBrowserCodeReceiver.ReceiveCodeAsync launches (and awaits, since it is modal) the WinForms dialog
1. WinForms dialog eventually closes (we'll assume it set the response token)
1. EmbeddedBrowserCodeReceiver.ReceiveCodeAsync sets the task result and exits.
1. AuthorizationCodeInstalledApp.AuthorizeAsync exchanges the token for an access token and saves the access token.
 * No further UI is involved for this process.
 * Native HTTP classes are used.

## Options to eliminate WinForms dependency
One option to eliminate the WinForms dependency is replace the call to the dialog with a hook. Since this code is executing a couple of layers internally, one option would be to have an event that the caller handles. The event arg properties would include the authorization URI and a response token. The handler would be expected to set the response token within the event handler.

Another way of doing this might be to make the receiver abstract and require a subclasse, which the user would define, to implement a LaunchBrowser method that returns the authorization response code.

Getting out a bigger hammer, another option would be to move the embedded browser code receiver completely, and put it in the samples repository. In that case, this pull request simply becomes adding the optional parameter to pass in a code receiver to AuthenticationCodeInstalledApp.
## Other thoughts
If we take the dialog form out of the project, but keep the receiver class, I would want to change its name because it would no longer reflect what it does. I would probably change it to BYOBCodeReceiver (i.e. Bring Your Own Browser). 

Is the concern about only Windows Forms, or is it about adding any kind of UI dependency to the project for the "installed app" platform? If I had used a WPF Window instead, would you have the same concern?

I must respectfully challenge the statement that Windows Forms is not something a lot of people are going to be using. While WPF is the shinier technology for new Windows applications, WinForms is still very much in use in the corporate environment. As Richard Campbell of .NET Rocks often says, "WinForms is not dead, it's just done." That is, It's not getting any new features, but it still works.

I will say that I don't think the other current alternatives for Windows-installed apps (LocalServerCodeReceiver and PromptCodeReceiver) are very good options. The prompt receiver really only works in a console app. LocalServerCodeReceiver has a nasty bug that hangs the app if the user closes the external browser (the receiver waits forever for a response that is never coming). The external browser has the problem that it cannot reliably close the browser window, which makes more work for the end user. When I tested with Chrome, I always had to manually close the browser tab after approving the request.

I would also make the case that this is not a burdensome dependency. It is part of the framework, so adding it imposes no new installation requirements. At runtime, the assemblies are only loaded into memory if somebody chooses to use this code receiver. For all other users, there will be no memory changes, other than those due to the new classes in the Platform Services DLL.